### PR TITLE
Convert `display-lazy-ads` to TypeScript

### PR DESCRIPTION
**Pull request by arelra**
## What does this change?

Convert `display-lazy-ads` to TypeScript

### CODE REVIEW

File: static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+
+const displayLazyAds = (): void => {
+	window.googletag?.pubads().collapseEmptyDivs();
+	window.googletag?.enableServices();
+
+	instantLoad();
+
+	dfpEnv.advertsToLoad.forEach((advert) => {
+		enableLazyLoad(advert);
+	});
```
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts:27]: Seems a bit odd to iterate over `dfpEnv.advertsToLoad` 3 separate times.

Do you find the following easier to parse? And then we get rid of `instantLoad` altogether.

```suggestion
	dfpEnv.advertsToLoad.forEach((advert) => {
	    if(advertsToInstantlyLoad.includes(advert.id))
	        loadAdvert(advert);
	    else
		    enableLazyLoad(advert);
	});
```

**Review by mxdvl — COMMENTED** [review]
I’m tempted to remove the `instantLoad` method.

File: static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts (hunk 1 above)
Comment (arelra) [reply static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts:27]: I agree there is some unneccessary looping here.

The only problem being it looks like line 12 is setting `dfpEnv.advertsToLoad`.

I'll take a look at simplifying thanks

Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts:27]: If you pass the index as the second param of `forEach`, you can do `dfpEnv.advertsToLoad.splice(i, 1)`.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/splice

Comment (mxdvl) [reply static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts:27]: Side effects are hard to reason about. Why do we need to modify `dfpEnv`? So many questions!

Comment (arelra) [reply static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts]: Yes working why this side effect is there is a larger task... in fact what all the state in `dfpEnv` is for at all.

I've refactored to reduce the loops from 4 to 3 for now.

File: static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+	array.reduce(
+		(result: [T[], T[]], element, i) => {
+			callback(element, i, array)
+				? result[0].push(element)
+				: result[1].push(element);
+
+			return result;
+		},
+		[[], []],
+	);
```
Comment (mxdvl) [inline static/src/javascripts/projects/commercial/modules/dfp/display-lazy-ads.ts:20]: This is what I was hoping for: a `partition` method.

How 👌 _Magnifique!_

Worth moving up to utils? Or at least exporting it?

**Review by mxdvl — APPROVED** [review]

File: static/src/javascripts/lib/partition.ts [hunk 3]
```diff
@@ -0,0 +1,13 @@
+const partition = <T>(
+	array: T[],
+	callback: (element: T, i: number, array: T[]) => boolean,
+): [T[], T[]] =>
+	array.reduce(
+		(result: [T[], T[]], e, i) => {
+			callback(e, i, array) ? result[0].push(e) : result[1].push(e);
```
Comment (mxdvl) [inline static/src/javascripts/lib/partition.ts:7]: FYI, [Typescript `lib.es5.d.ts` will refer to this callback as the `predicate`](https://github.com/microsoft/TypeScript/blob/4ce902c5fa31d402be78709b3c44bc60405bad02/lib/lib.es5.d.ts#L4070)

Comment (arelra) [reply static/src/javascripts/lib/partition.ts:7]: You're right I should call it 'predicate'. I did in the tests but not the impl!

Comment (mxdvl) [reply static/src/javascripts/lib/partition.ts:7]: [MDN calls it a `callbackFn`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter#syntax), so you’re not wrong ;)

**Comment by mxdvl** [comment]
[Lovely!](https://guardian-commercial-bundle.netlify.app/)

partition graph

**Pull request merged**